File: python/503_下一个更大元素II_medium.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ret, f = [-1]*n, []
        for i in range(2*n-1):
            while f and nums[f[-1]] < nums[i%n]:
                ret[f[-1]] = nums[i%n]
                f.pop()
            f.append(i%n)
        return ret

File: cpp/503_下一个更大元素II_medium.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        vector<int> ret(n, -1);
        stack<int> f;
        for (int i=0; i<2*n-1; i++) {
            while (!f.empty() && nums[f.top()] < nums[i%n]) {
                ret[f.top()] = nums[i%n];
                f.pop();
            }
            f.push(i%n);
        }
        return ret;
    }
};